types/query/filtered-pagination.go

```diff
@@ -0,0 +1,148 @@
+package query
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// defaultLimit is the default `limit` for queries
+// if the `limit` is not supplied, FilteredPaginate will use `defaultLimit`
+const defaultLimit = 100
+
+// Example Code
+/*
+func (...) GetProposalsFiltered(ctx sdk.Context, params types.QueryProposalsParams) types.Proposals {
+	var filteredProposals []types.Proposal
+
+	pageReq := &query.PageRequest{Key: nil, Limit: 10, Offset: 10, CountTotal: true}
+	res, err := query.FilteredPaginate(accountStore, request.Req, func(key []byte, value []byte, isValid bool) (hit, error) {
+		var p types.Proposal
+		err := app.Codec().UnmarshalBinaryBare(value, &p)
+		if err != nil {
+			return err
+		}
+
+		matchVoter, matchDepositor, matchStatus := true, true, true
+
+		// match status (if supplied/valid)
+		if types.ValidProposalStatus(params.ProposalStatus) {
+			matchStatus = p.Status == params.ProposalStatus
+		}
+
+		// match voter address (if supplied)
+		if len(params.Voter) > 0 {
+			_, matchVoter = keeper.GetVote(ctx, p.ProposalID, params.Voter)
+		}
+
+		// match depositor (if supplied)
+		if len(params.Depositor) > 0 {
+			_, matchDepositor = keeper.GetDeposit(ctx, p.ProposalID, params.Depositor)
+		}
+
+		if matchVoter && matchDepositor && matchStatus {
+			if isValid {
+				filteredProposals = append(filteredProposals, p)
+			}
```

Comment:
We can eliminate `isValid` usage but we need to split the results when using `offset`. i.e.,:

```go
if req.Offset {
    return types.QueryProposalsResponse{Proposals: filteredProposals[offset:], Res: res}
}

return types.QueryProposalsResponse{Proposals: filteredProposals, Res: res}
```

Reply:
This looks odd as the callee needs to handle the results slicing, which should be part of pagination.